server/src/fstar_messages.ts

```diff
@@ -0,0 +1,359 @@
+////////////////////////////////////////////////////////////////////////////////////
+
+// ProtocolInfo: The first response from fstar.exe when it is first launched
+export interface ProtocolInfo {
+	kind: 'protocol-info';
+	version: number;
+	features: string[];
+}
+
+export function isProtocolInfo(object: any): boolean {
+	return object
+		&& object.kind && object.kind === 'protocol-info'
+		&& object.version
+		&& object.features;
+}
+
+export function parseProtocolInfo(object: any): ProtocolInfo | undefined {
+	if (isProtocolInfo(object)) return object;
+}
+
+// An FStarRange is a range of positions in some source file
+// A position is a line number and a column number
+// A quirk is that the line number is 1-based, but the column number is 0-based
+// In contrast, the LSP protocol uses 0-based line and column numbers
+// So, we need to convert between the two
+export interface FStarRange {
+	fname: string;
+	beg: number[];
+	end: number[]
+}
+
+export function isFStarRange(object: any): boolean {
+	return object
+		&& object.fname
+		&& object.beg
+		&& object.end;
+}
+
+export function parseFStarRange(object: any): FStarRange | undefined {
+	if (isFStarRange(object)) return object;
+}
+
+// IdeSymbol: This message is sent from fstar.exe in response to a
+// request for a onHover or onDefnition symbol lookup
+export interface IdeSymbol {
+	kind: 'symbol';
+	name: string;
+	type: string;
+	documentation: string;
+	definition: string;
+	"defined-at": FStarRange;
+	"symbol-range": FStarRange;
+	symbol: string;
+}
+
+export function isIdeSymbol(object: any): boolean {
+	return object
+		&& object.kind && object.kind === 'symbol'
+		&& object.name
+		&& object.type
+		&& object.documentation
+		&& object.definition
+		&& object["defined-at"] && isFStarRange(object["defined-at"])
+		&& object["symbol-range"] && isFStarRange(object["symbol-range"])
+		&& object.symbol;
+}
+
+export function parseIdeSymbol(object: any): IdeSymbol | undefined {
+	if (isIdeSymbol(object)) return object;
+}
+
+// IdeProofState: fstar.exe sends informative messages when running tactics
+// The server does not explicitly request the proof state---these messages
+// are just sent by fstar.exe as a side-effect of running tactics
+// The server stores the proof state in a table, and uses it to display
+// the proof state in the onHover message
+export interface IdeProofState {
+	label: string; // User-provided label, e.g., dump "A"
+	depth: number; // The depth of this dump message (not displayed)
+	urgency: number; // Urgency (not displayed)
+	goals: IdeProofStateContextualGoal[]; // The main proof state
+	"smt-goals": IdeProofStateContextualGoal[]; // SMT goals
+	location: FStarRange; // The location of the tactic that triggered a proof state dump
+}
+
+export function isIdeProofState(object: any): boolean {
+	return object
+		&& object.label
+		&& object.depth
+		&& object.urgency
+		// TODO(klinvill): How rigorous should these checks be? There's a
```

Comment:
AFAICT this particular function isn't even used.

The generic answer is IMHO: we assume that F* follows the protocol and only sends well-formed responses.  So, if the request with corresponding the query id is, say, a `lookup`, then the response will be an `IdeSymbol` (and we don't have to check anything except the query id).

Reply:
Also related to #20.